File: apps/api/app/schemas/workspaces.py
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.schemas.currency import SupportedCurrency


WorkspaceType = Literal["personal", "team"]
WorkspaceRole = Literal["owner", "admin", "member", "viewer"]
AssignableMemberRole = Literal["admin", "member", "viewer"]


class Workspace(BaseModel):
    id: UUID
    type: WorkspaceType
    name: str
    role: WorkspaceRole
    currency: SupportedCurrency = "SAR"
    auto_delete_after_extraction: bool = False
    currency_locked: bool = False
    member_count: int | None = None

    model_config = ConfigDict(from_attributes=True)


class WorkspacesListResponse(BaseModel):
    workspaces: list[Workspace]


class WorkspaceCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=120)


class WorkspaceUpdateRequest(BaseModel):
    auto_delete_after_extraction: bool | None = None
    currency: SupportedCurrency | None = None

    model_config = ConfigDict(extra="forbid")

    @model_validator(mode="after")
    def require_update_field(self) -> "WorkspaceUpdateRequest":
        if not any(
            field in self.model_fields_set and getattr(self, field) is not None
            for field in ("auto_delete_after_extraction", "currency")
        ):
            raise ValueError("At least one field is required.")
        return self


class WorkspaceSettingsResponse(BaseModel):
    id: UUID
    currency: SupportedCurrency = "SAR"
    auto_delete_after_extraction: bool


class WorkspaceMember(BaseModel):
    user_id: UUID
    email: str
    role: WorkspaceRole


class WorkspaceMembersListResponse(BaseModel):
    members: list[WorkspaceMember]


class MemberAddRequest(BaseModel):
    email: str
    role: str


class MemberRoleUpdateRequest(BaseModel):
    role: str

File: apps/web/lib/api/workspaces.ts
import { apiFetch } from "./client";
import type { SupportedCurrency } from "../currency";

export type WorkspaceRole = "owner" | "admin" | "member" | "viewer";
export type WorkspaceType = "personal" | "team";

export type WorkspaceSummary = {
  id: string;
  type: WorkspaceType;
  name: string;
  role: WorkspaceRole;
  currency: SupportedCurrency;
  auto_delete_after_extraction: boolean;
  currency_locked: boolean;
};

export type WorkspaceDetail = WorkspaceSummary & {
  member_count: number;
};

export type WorkspaceSettingsResponse = {
  id: string;
  currency: SupportedCurrency;
  auto_delete_after_extraction: boolean;
};

export async function getWorkspaces() {
  return apiFetch<{ workspaces: WorkspaceSummary[] }>("/workspaces");
}

export async function getWorkspace(workspaceId: string) {
  return apiFetch<WorkspaceDetail>(`/workspaces/${workspaceId}`);
}

export async function createWorkspace(name: string) {
  return apiFetch<WorkspaceSummary>("/workspaces", {
    method: "POST",
    body: JSON.stringify({ name }),
  });
}

export async function updateWorkspaceAutoDelete(workspaceId: string, autoDeleteAfterExtraction: boolean) {
  return apiFetch<WorkspaceSettingsResponse>(`/workspaces/${workspaceId}`, {
    method: "PATCH",
    body: JSON.stringify({ auto_delete_after_extraction: autoDeleteAfterExtraction }),
  });
}

export async function updateWorkspaceCurrency(workspaceId: string, currency: SupportedCurrency) {
  return apiFetch<WorkspaceSettingsResponse>(`/workspaces/${workspaceId}`, {
    method: "PATCH",
    body: JSON.stringify({ currency }),
  });
}
